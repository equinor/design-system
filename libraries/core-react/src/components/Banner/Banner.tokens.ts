/* eslint-disable camelcase */
import { tokens } from '@equinor/eds-tokens'
import type { ComponentToken } from '@equinor/eds-tokens'

const {
  typography: {
    paragraph: { body_long: messageTypography },
  },
  spacings: {
    comfortable: { medium },
  },
  colors: {
    infographic: {
      primary__moss_green_13: { rgba: infoBackground },
      primary__energy_red_13: { rgba: warningBackground },
      primary__moss_green_100: { rgba: infoColor },
      primary__energy_red_100: { rgba: warningColor },
    },
    ui: {
      background__default: { rgba: backgroundColor },
    },
  },
  shape: {
    circle: { minHeight, minWidth, borderRadius },
  },
} = tokens

export type BannerToken = ComponentToken & {
  icon: {
    info: {
      background: string
      color: string
    }
    warning: {
      background: string
      color: string
    }
    shape: {
      minHeight: string
      minWidth: string
      borderRadius: string
    }
  }
}

export const banner: BannerToken = {
  typography: {
    ...messageTypography,
  },
  background: backgroundColor,
  spacings: {
    left: medium,
    right: medium,
    top: medium,
    bottom: medium,
  },
  icon: {
    info: {
      background: infoBackground,
      color: infoColor,
    },
    warning: {
      background: warningBackground,
      color: warningColor,
    },
    shape: {
      minHeight,
      minWidth,
      borderRadius,
    },
  },
}
