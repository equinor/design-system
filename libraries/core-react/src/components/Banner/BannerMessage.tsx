import * as React from 'react'
import { FC } from 'react'
import styled from 'styled-components'
import { Typography } from '../Typography'
import { TypographyProps } from '../Typography/Typography'

const StyledBannerMessage = styled(Typography)``

export type BannerMessageProps = {
  /** Text content */
  children: string
} & Omit<TypographyProps, 'children'>

export const BannerMessage: FC<BannerMessageProps> = ({
  children,
  ...props
}) => {
  return (
    <StyledBannerMessage variant="body_long" {...props}>
      {children}
    </StyledBannerMessage>
  )
}

// BannerMessage.displayName = 'eds-banner-message'
