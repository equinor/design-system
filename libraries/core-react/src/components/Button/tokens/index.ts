import { ButtonTokenSet } from '../Button.types'

import * as contained from './contained'
import * as outlined from './outlined'
import * as ghost from './ghost'
import * as icon from './icon'

type ButtonTokens = {
  primary: ButtonTokenSet
  secondary: ButtonTokenSet
  danger: ButtonTokenSet
}

export const token: ButtonTokens = {
  primary: {
    contained: contained.primary,
    outlined: outlined.primary,
    ghost: ghost.primary,
    ghost_icon: icon.primary,
  },
  secondary: {
    contained: contained.secondary,
    outlined: outlined.secondary,
    ghost: ghost.secondary,
    ghost_icon: icon.secondary,
  },
  danger: {
    contained: contained.danger,
    outlined: outlined.danger,
    ghost: ghost.danger,
    ghost_icon: icon.danger,
  },
}
