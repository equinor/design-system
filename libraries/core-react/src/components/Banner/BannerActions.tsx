import { forwardRef, HTMLAttributes, ReactNode } from 'react'
import styled from 'styled-components'
import * as tokens from './Banner.tokens'

const { enabled } = tokens

type BannerActionsPlacement = 'bottom' | 'left'

type StyledBannerActionsProps = {
  /** Where to place the actions */
  placement: BannerActionsPlacement
}

const StyledBannerActions = styled.div<StyledBannerActionsProps>`
  display: flex;
  margin-left: ${enabled.spacings.left};
  grid-gap: 8px;
  grid-column: ${({ placement }) => (placement === 'bottom' ? '1/-1' : 'auto')};
  ${({ placement }) =>
    placement === 'bottom' && {
      marginTop: enabled.spacings.top,
      marginLeft: '0',
    }}
`

export type BannerActionsProps = {
  children: ReactNode
  placement?: BannerActionsPlacement
} & HTMLAttributes<HTMLDivElement>

export const BannerActions = forwardRef<HTMLDivElement, BannerActionsProps>(
  function BannerActions({ children, placement = 'left', ...rest }, ref) {
    const props = {
      ref,
      ...rest,
    }
    return (
      <StyledBannerActions placement={placement} {...props}>
        {children}
      </StyledBannerActions>
    )
  },
)
