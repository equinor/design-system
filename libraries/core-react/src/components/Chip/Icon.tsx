import styled, { css } from 'styled-components'
import { close } from '@equinor/eds-icons'
import { Icon as Icon_ } from '../Icon'
import * as tokens from './Chip.tokens'
import { bordersTemplate } from '@utils'

Icon_.add({ close })

const { primary, error } = tokens

type IconProps = {
  variant: 'active' | 'error' | 'default'
  disabled: boolean
}

export const Icon = styled(Icon_)<IconProps>`
  cursor: pointer;
  padding: 1px;
  ${bordersTemplate(primary.entities.icon.border)}

  z-index: 11;

  &:hover {
    ${({ variant }) => {
      switch (variant) {
        case 'error':
          return css`
            background: ${error.entities.icon.background};
            svg {
              fill: ${error.entities.icon.typography.color};
            }
          `
        default:
          return css`
            background: ${primary.entities.icon.states.hover.background};
            svg {
              fill: ${primary.states.hover.typography.color};
            }
          `
      }
    }};
  }

  ${({ disabled }) =>
    disabled &&
    css`
      cursor: not-allowed;
      &:hover {
        background: transparent;
      }
    `}
`
