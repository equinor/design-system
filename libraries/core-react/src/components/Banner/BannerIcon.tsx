import * as React from 'react'
import { FC, HTMLAttributes, ReactNode } from 'react'
import styled from 'styled-components'
import * as tokens from './Banner.tokens'
import { Icon } from '../Icon'
import { bordersTemplate } from '@utils'

type BannerIconVariant = 'info' | 'warning'

type StyledBannerIconProps = {
  variant: BannerIconVariant
}

const { enabled, info, warning } = tokens

const StyledBannerIcon = styled.span<StyledBannerIconProps>`
  display: inline-flex;
  align-items: center;
  justify-content: center;
  flex-shrink: 0;
  ${bordersTemplate(enabled.entities.icon.border)};
  background-color: ${({ variant }) =>
    variant === 'warning'
      ? warning.entities.icon.background
      : info.entities.icon.background};
  width: ${enabled.entities.icon.width};
  height: ${enabled.entities.icon.height};
  margin-right: ${enabled.spacings.right};
`

type BannerIconProps = {
  /** Which icon background and fill color to use. Info = green, warning = red */
  variant?: BannerIconVariant
  /** @ignore */
  children: ReactNode
} & HTMLAttributes<HTMLSpanElement>

export const BannerIcon: FC<BannerIconProps> = ({
  children,
  variant = 'info',
  ...props
}) => {
  const childrenWithColor = React.Children.map(children, (child) => {
    const color =
      variant === 'warning'
        ? warning.entities.icon.typography.color
        : info.entities.icon.typography.color
    return (
      (React.isValidElement(child) &&
        child.type === Icon &&
        React.cloneElement(child, {
          color,
        })) ||
      child
    )
  })
  return (
    <StyledBannerIcon variant={variant} {...props}>
      {childrenWithColor}
    </StyledBannerIcon>
  )
}
