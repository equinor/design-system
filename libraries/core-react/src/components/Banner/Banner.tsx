import * as React from 'react'
import { FC, HTMLAttributes, ReactNode } from 'react'
import styled from 'styled-components'
import * as tokens from './Banner.tokens'
import { Divider } from '../Divider'
import { BannerIcon } from './BannerIcon'
import { spacingsTemplate } from '../../utils'

const StyledBanner = styled.div``

type ContentProps = {
  hasIcon: boolean
}

const { enabled } = tokens

const Content = styled.div<ContentProps>`
  ${spacingsTemplate(enabled.spacings)}

  display: grid;
  grid-template-columns: ${({ hasIcon }) =>
    hasIcon ? 'min-content 1fr auto' : '1fr auto'};
  align-items: center;
  background-color: ${enabled.background};
`

const NonMarginDivider = styled(Divider)`
  margin: 0;
  height: 2px;
`

export type BannerProps = {
  children: ReactNode
} & HTMLAttributes<HTMLDivElement>

export const Banner: FC<BannerProps> = ({ children, className, ...props }) => {
  const childrenWhereBannerIcon: boolean[] = React.Children.map(
    children,
    (child) => {
      return React.isValidElement(child) && child.type === BannerIcon
    },
  )
  const hasIcon = childrenWhereBannerIcon.some((bool) => bool)

  return (
    <StyledBanner {...props} className={className} role="alert">
      <Content hasIcon={hasIcon}>{children}</Content>
      <NonMarginDivider color="light" />
    </StyledBanner>
  )
}
