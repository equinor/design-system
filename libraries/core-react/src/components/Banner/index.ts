import { Banner as BaseBanner, BannerProps } from './Banner'
import { BannerIcon, BannerIconProps } from './BannerIcon'
import { BannerMessage, BannerMessageProps } from './BannerMessage'
import { BannerActions, BannerActionsProps } from './BannerActions'

type BannerCompoundProps = typeof BaseBanner & {
  Icon: typeof BannerIcon
  Message: typeof BannerMessage
  Actions: typeof BannerActions
}

const Banner = BaseBanner as BannerCompoundProps
Banner.Icon = BannerIcon
Banner.Message = BannerMessage
Banner.Actions = BannerActions

Banner.Icon.displayName = 'Banner.Icon'
Banner.Message.displayName = 'Banner.Message'
Banner.Actions.displayName = 'Banner.Actions'

export { Banner }
export type {
  BannerProps,
  BannerMessageProps,
  BannerIconProps,
  BannerActionsProps,
}
