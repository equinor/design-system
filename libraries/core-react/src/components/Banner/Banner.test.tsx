/* eslint-disable no-undef */
import * as React from 'react'
import { render, cleanup } from '@testing-library/react'
import '@testing-library/jest-dom'
import 'jest-styled-components'
import styled from 'styled-components'
import { add } from '@equinor/eds-icons'
import { Banner } from '.'
import { Icon } from '../Icon'
import * as tokens from './Banner.tokens'

const { BannerMessage, BannerIcon, BannerActions } = Banner
Icon.add({ add })

const { info, warning } = tokens

const StyledBanner = styled(Banner)`
  position: relative;
`

const rgbaTrim = (x: string) => x.split(' ').join('')

afterEach(cleanup)

describe('Banner', () => {
  it('Can extend the css for the component', () => {
    const { container } = render(
      <StyledBanner>
        <BannerMessage>styled banner</BannerMessage>
      </StyledBanner>,
    )
    expect(container.firstChild).toHaveStyleRule('position', 'relative')
  })
  it('Has provided Message', () => {
    const bannerText = 'Banner test message'
    const { queryByText } = render(
      <Banner>
        <BannerMessage>{bannerText}</BannerMessage>
      </Banner>,
    )
    expect(queryByText(bannerText)).toBeDefined()
  })
  it('Has provided Icon', () => {
    const bannerText = 'Banner test'
    const iconTestId = 'banner-icon-test'
    const { queryByText, queryByTestId } = render(
      <Banner>
        <BannerIcon>
          <Icon name="add" data-testid={iconTestId} />
        </BannerIcon>
        <BannerMessage>{bannerText}</BannerMessage>
      </Banner>,
    )
    expect(queryByText(bannerText)).toBeDefined()
    expect(queryByTestId(iconTestId)).toBeDefined()
  })
  it('Has provided Actions', () => {
    const bannerText = 'Banner test'
    const actionButtonText = 'Banner action button text'
    const { queryByText } = render(
      <Banner>
        <BannerMessage>{bannerText}</BannerMessage>
        <BannerActions>
          <button type="button">{actionButtonText}</button>
        </BannerActions>
      </Banner>,
    )
    expect(queryByText(bannerText)).toBeDefined()
    expect(queryByText(actionButtonText)).toBeDefined()
  })

  it('Has correct default icon styling as info color', () => {
    const bannerText = 'Banner test'
    const iconWrapperTestId = 'banner-icon-wrapper-test'
    const { queryByTestId, container } = render(
      <Banner>
        <BannerIcon data-testid={iconWrapperTestId}>
          <Icon name="add" />
        </BannerIcon>
        <BannerMessage>{bannerText}</BannerMessage>
      </Banner>,
    )
    const iconSvg = container.querySelector('svg')
    expect(queryByTestId(iconWrapperTestId)).toHaveStyleRule(
      'background-color',
      rgbaTrim(info.entities.icon.background),
    )
    expect(iconSvg).toHaveAttribute('fill', info.entities.icon.typography.color)
  })
  it('Has correct warning icon styling', () => {
    const bannerText = 'Banner test'
    const iconWrapperTestId = 'banner-icon-wrapper-test'
    const { queryByTestId, container } = render(
      <Banner>
        <BannerIcon variant="warning" data-testid={iconWrapperTestId}>
          <Icon name="add" />
        </BannerIcon>
        <BannerMessage>{bannerText}</BannerMessage>
      </Banner>,
    )
    const iconSvg = container.querySelector('svg')
    expect(queryByTestId(iconWrapperTestId)).toHaveStyleRule(
      'background-color',
      rgbaTrim(warning.entities.icon.background),
    )
    expect(iconSvg).toHaveAttribute(
      'fill',
      warning.entities.icon.typography.color,
    )
  })
})
