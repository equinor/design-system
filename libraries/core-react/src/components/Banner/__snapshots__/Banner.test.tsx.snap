// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banner Matches snapshot 1`] = `
<DocumentFragment>
  .c3 {
  border: none;
  background-color: var(--eds_ui_background__light,rgba(247,247,247,1));
  margin-top: 16px;
  margin-bottom: 15px;
  height: 1px;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  border-radius: 100px;
  background-color: var(--eds_infographic_primary__moss_green_13,rgba(222,237,238,1));
  width: 40px;
  height: 40px;
  margin-right: 16px;
}

.c0 {
  padding-left: 16px;
  padding-top: 16px;
  padding-right: 16px;
  padding-bottom: 16px;
  display: grid;
  grid-template-columns: min-content 1fr auto;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: var(--eds_ui_background__default,rgba(255,255,255,1));
}

.c4 {
  margin: 0;
  height: 2px;
}

.c2 {
  margin: 0;
  color: rgba(61,61,61,1);
  font-family: Equinor;
  font-size: 1.000rem;
  font-weight: 400;
  line-height: 1.500em;
  text-align: left;
}

<div
    class="Banner__StyledBanner-sc-1ju451i-0"
    role="alert"
  >
    <div
      class="c0"
    >
      <span
        class="c1"
      >
        <svg
          aria-hidden="true"
          class="Icon__StyledSvg-sc-6evbi1-0"
          fill="var(--eds_infographic_primary__moss_green_100, rgba(0, 112, 121, 1))"
          height="24px"
          viewBox="0 0 24 24"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            class="Icon__StyledPath-sc-6evbi1-1"
            clip-rule="evenodd"
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            data-testid="eds-icon-path"
            fill-rule="evenodd"
            height="24"
          />
        </svg>
      </span>
      <p
        class="c2 BannerMessage__StyledBannerMessage-sc-1lfqos1-0"
      >
        Banner message
      </p>
    </div>
    <hr
      class="c3 c4"
    />
  </div>
</DocumentFragment>
`;
